bpf/bpf_lxc.c

```diff
@@ -1003,12 +1008,30 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx, __u32 *d
 	if (!revalidate_data(ctx, &data, &data_end, &ip4))
 		return DROP_INVALID;
 
+#if defined(ENABLE_HOST_FIREWALL) && !defined(ENABLE_ROUTING)
+	/* If the destination is the local host and per-endpoint routes are
+	 * enabled, jump to the bpf_host program to enforce ingress host policies.
+	 */
+	if (*dst_id == HOST_ID) {
+		ctx_store_meta(ctx, CB_FROM_HOST, 0);
+		tail_call_static(ctx, &POLICY_CALL_MAP, HOST_EP_ID);
+		return DROP_MISSED_TAIL_CALL;
+	}
+#endif /* ENABLE_HOST_FIREWALL && !ENABLE_ROUTING */
+
 	/* Allow a hairpin packet to be redirected even if ENABLE_ROUTING is
-	 * disabled. Otherwise, the packet will be dropped by the kernel if
-	 * it is going to be routed via an interface it came from after it has
-	 * been passed to the stack.
+	 * disabled (e.g. the case in per-endpoint routes). Otherwise, the
+	 * packet will be dropped by the kernel if it is going to be routed via
+	 * an interface it came from after it has been passed to the stack.
+	 *
+	 * For the case where ENABLE_ROUTING is disabled, but the fast redirect
+	 * enabled we do lookup the local endpoint here to check whether we
+	 * must pass the packet up the stack for the host itself. We also want
+	 * to run through ipv4_local_delivery to enforce ingress policies for
+	 * that endpoint.
```

Comment:
Who is "we" here in both instances? It's unclear which component(s) are acting.

```suggestion
	 * If ENABLE_ROUTING is disabled but fast redirect is
	 * enabled, we do lookup the local endpoint here to check whether we
	 * must pass the packet up the stack for the host itself. We also want
	 * to run through `ipv4_local_delivery` to enforce ingress policies for
	 * that endpoint.
```